File: scripts/pyrite_kinetics.py
#!/usr/bin/env python3
"""Simulate time-dependent pyrite oxidation (acid mine drainage generation).

Reads one JSON request on stdin, writes exactly one line of JSON to stdout.

This is the KINETIC counterpart to `phreeqc_run.py`. Static ABA screening
(MPA/NAPP) says how much acid a rock *could* make; equilibrium speciation says
what a given water *is*. Neither says how fast acid appears, which is the
question that decides whether a pit needs treatment in month 3 or year 30.

Rate law: Williamson & Rimstidt (1994), as shipped in the RATES block of the
database. In the repaired WATEQ4F database it evaluates as

    log10(rate) = -10.19 + parm1 + parm3*log(mO2) + parm4*log(mH+)
                  + parm2*log10(m/m0)

    parm1 = log10(A/V) with A/V in 1/dm      parm2 = exponent on (m/m0)
    parm3 = exponent on O2 (0.5)             parm4 = exponent on H+ (-0.11)

WARNING ON DATABASE PORTABILITY: `phreeqc.dat` implements the same reference
with a different intercept (-8.19) and treats parm1 as log10(specific area in
m2/mol) combined with log10(M0). The SAME `-parms` values therefore mean
DIFFERENT things in the two databases. This script pins the database and
reports which one produced the numbers.

Request:

    {"pyrite_mol_per_kgw": 0.05,
     "initial_ph": 6.5,
     "initial_o2_mmol": 0.27,
     "replenish_o2": true,
     "o2_partial_pressure_log10": -0.68,
     "steps_days": [1, 30, 90, 180, 365],
     "parms": [1.0, 0.67, 0.5, -0.11],
     "temperature_c": 25.0,
     "neutralising_phases": ["Calcite"]}

Contract: {"status":"ok","database":...,"series":[...],"guards":{...}}

Exit codes: 0 ok, 1 runtime error, 2 invalid request.
"""

from __future__ import annotations

import argparse
import hashlib
import json
import os
import sys
import traceback
from pathlib import Path

INVALID_REQUEST = 2
RUNTIME_ERROR = 1

DEFAULT_DATABASE = "wateq4f_PWN_repaired.dat"
DEFAULT_DATABASE_DIR = Path(__file__).resolve().parent.parent / "resources" / "phreeqc"

# Williamson & Rimstidt (1994) exponents: rate ∝ mO2^0.5 * mH+^-0.11
DEFAULT_PARMS = [1.0, 0.67, 0.5, -0.11]

SECONDS_PER_DAY = 86400.0
MAX_STEPS = 200

# FeS2 releases 2 mol S per mol Fe. A measured ratio far from 2 means Fe was
# removed by secondary precipitation, so dissolved Fe understates oxidation.
STOICHIOMETRIC_S_PER_FE = 2.0
STOICHIOMETRY_TOLERANCE = 0.15

# Below this pH change across the final half of the series the reaction has
# effectively stalled.
STALL_PH_DELTA = 0.02

# Above this fraction of pyrite consumed the run is depletion-limited.
DEPLETION_FRACTION = 0.99


def emit(payload: dict) -> None:
    sys.stdout.write(json.dumps(payload, separators=(",", ":")) + "\n")
    sys.stdout.flush()


def fail_invalid(message: str):
    emit({"status": "invalid_request", "error": message})
    sys.exit(INVALID_REQUEST)


def fail_error(message: str):
    emit({"status": "error", "error": message})
    sys.exit(RUNTIME_ERROR)


def sha256_of(path: Path) -> str:
    digest = hashlib.sha256()
    with open(path, "rb") as handle:
        for chunk in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(chunk)
    return digest.hexdigest()


def finite(value, default: float = 0.0) -> float:
    try:
        number = float(value)
    except (TypeError, ValueError):
        return default
    if number != number or number in (float("inf"), float("-inf")):
        return default
    return number


def validate(request: dict) -> dict:
    pyrite = finite(request.get("pyrite_mol_per_kgw"), float("nan"))
    if not pyrite > 0.0:
        fail_invalid("pyrite_mol_per_kgw must be a positive finite number")

    initial_ph = finite(request.get("initial_ph"), 6.5)
    if not 0.0 < initial_ph < 14.0:
        fail_invalid("initial_ph must be between 0 and 14")

    o2 = finite(request.get("initial_o2_mmol"), 0.27)
    if o2 < 0.0:
        fail_invalid("initial_o2_mmol must not be negative")

    temperature = finite(request.get("temperature_c"), 25.0)
    if not -10.0 <= temperature <= 100.0:
        fail_invalid("temperature_c must be between -10 and 100")

    steps = request.get("steps_days") or [1, 30, 90, 180, 365]
    if not isinstance(steps, list) or not steps:
        fail_invalid("steps_days must be a non-empty list")
    if len(steps) > MAX_STEPS:
        fail_invalid("steps_days must contain at most {} entries".format(MAX_STEPS))
    parsed_steps = []
    previous = 0.0
    for index, step in enumerate(steps):
        value = finite(step, float("nan"))
        if not value > 0.0:
            fail_invalid("steps_days[{}] must be positive".format(index))
        if value <= previous:
            fail_invalid("steps_days must be strictly increasing (cumulative times)")
        previous = value
        parsed_steps.append(value)

    parms = request.get("parms") or DEFAULT_PARMS
    if not isinstance(parms, list) or len(parms) != 4:
        fail_invalid("parms must be a list of exactly 4 numbers")
    parms = [finite(p, float("nan")) for p in parms]
    if any(p != p for p in parms):
        fail_invalid("parms must contain only finite numbers")

    phases = request.get("neutralising_phases") or []
    if not isinstance(phases, list) or any(not isinstance(p, str) for p in phases):
        fail_invalid("neutralising_phases must be a list of phase names")

    database = str(request.get("database", DEFAULT_DATABASE))
    if "/" in database or "\\" in database or database.startswith("."):
        fail_invalid("database must be a bare file name inside the resources directory")

    return {
        "pyrite_mol_per_kgw": pyrite,
        "initial_ph": initial_ph,
        "initial_o2_mmol": o2,
        "replenish_o2": bool(request.get("replenish_o2", True)),
        "o2_partial_pressure_log10": finite(request.get("o2_partial_pressure_log10"), -0.68),
        "steps_days": parsed_steps,
        "parms": parms,
        "temperature_c": temperature,
        "neutralising_phases": phases,
        "database": database,
        "database_directory": Path(request.get("database_directory", DEFAULT_DATABASE_DIR)),
    }


def build_input(spec: dict) -> str:
    blocks = []
    blocks.append(
        "SOLUTION 1\n"
        "    units   mmol/kgw\n"
        "    temp    {temp}\n"
        "    pH      {ph}\n"
        "    O(0)    {o2}\n".format(
            temp=spec["temperature_c"],
            ph=spec["initial_ph"],
            o2=spec["initial_o2_mmol"],
        )
    )

    equilibrium = []
    if spec["replenish_o2"]:
        # A large O2(g) reservoir at fixed partial pressure: an open pit or waste
        # dump exposed to the atmosphere.
        equilibrium.append(
            "    O2(g)  {:.4f}  1000.0".format(spec["o2_partial_pressure_log10"])
        )
    for phase in spec["neutralising_phases"]:
        equilibrium.append("    {}  0.0  10.0".format(phase))
    if equilibrium:
        blocks.append("EQUILIBRIUM_PHASES 1\n" + "\n".join(equilibrium) + "\n")

    step_seconds = " ".join(
        "{:.6f}".format(day * SECONDS_PER_DAY) for day in spec["steps_days"]
    )
    blocks.append(
        "SELECTED_OUTPUT\n"
        "    -reset      false\n"
        "    -time       true\n"
        "    -ph         true\n"
        "    -pe         true\n"
        "    -totals     Fe S(6)\n"
        "    -kinetic_reactants Pyrite\n"
    )
    blocks.append(
        "KINETICS 1\n"
        "Pyrite\n"
        "    -formula  FeS2  1.0\n"
        "    -m0       {m0}\n"
        "    -parms    {p0} {p1} {p2} {p3}\n"
        "    -tol      1e-10\n"
        "    -steps    {steps}\n"
        "END\n".format(
            m0=spec["pyrite_mol_per_kgw"],
            p0=spec["parms"][0],
            p1=spec["parms"][1],
            p2=spec["parms"][2],
            p3=spec["parms"][3],
            steps=step_seconds,
        )
    )
    return "\n".join(blocks)


def run(request: dict) -> dict:
    from phreeqpython.viphreeqc import VIPhreeqc

    spec = validate(request)
    database_path = spec["database_directory"] / spec["database"]
    if not database_path.is_file():
        fail_invalid("database not found: {}".format(database_path))

    # A fresh interpreter per run: VIPhreeqc retains state between run_string
    # calls, so a reused instance leaks the previous run's kinetic amounts into
    # the first reported row.
    ip = VIPhreeqc()
    ip.load_database(database_path)
    ip.run_string(build_input(spec))

    columns = [ip.get_selected_output_value(0, c) for c in range(ip.column_count)]
    index = {name: position for position, name in enumerate(columns)}
    for required in ("time", "pH", "Fe(mol/kgw)", "S(6)(mol/kgw)"):
        if required not in index:
            fail_error("selected output missing column {}".format(required))
    pyrite_column = index.get("k_Pyrite")

    series = []
    for row in range(1, ip.row_count):
        values = [ip.get_selected_output_value(row, c) for c in range(ip.column_count)]
        time_seconds = finite(values[index["time"]], -1.0)
        # PHREEQC writes the pre-kinetics solution with time = -99; its kinetic
        # column is not yet meaningful for this run.
        if time_seconds < 0.0:
            continue
        remaining = (
            finite(values[pyrite_column], 0.0) if pyrite_column is not None else None
        )
        series.append(
            {
                "time_days": time_seconds / SECONDS_PER_DAY,
                "ph": finite(values[index["pH"]]),
                "pe": finite(values[index["pe"]]) if "pe" in index else None,
                "fe_mol_kgw": finite(values[index["Fe(mol/kgw)"]]),
                "sulfate_mol_kgw": finite(values[index["S(6)(mol/kgw)"]]),
                "pyrite_remaining_mol_kgw": remaining,
            }
        )

    if not series:
        fail_error("kinetic run produced no time steps")

    return {
        "status": "ok",
        "database": spec["database"],
        "database_sha256": sha256_of(database_path),
        "rate_law": "Williamson & Rimstidt (1994) pyrite oxidation, RATES block of the loaded database",
        "parms": spec["parms"],
        "replenish_o2": spec["replenish_o2"],
        "neutralising_phases": spec["neutralising_phases"],
        "initial_pyrite_mol_kgw": spec["pyrite_mol_per_kgw"],
        "series": series,
        "guards": build_guards(spec, series),
    }


def build_guards(spec: dict, series: list) -> dict:
    """Compute the honesty guards for a kinetic pyrite run."""
    final = series[-1]
    first = series[0]

    initial_pyrite = spec["pyrite_mol_per_kgw"]
    remaining = final.get("pyrite_remaining_mol_kgw")
    consumed_fraction = None
    if remaining is not None and initial_pyrite > 0.0:
        consumed_fraction = max(0.0, (initial_pyrite - remaining) / initial_pyrite)

    # Stall detection: compare the pH drop across the second half of the series.
    midpoint = series[len(series) // 2]
    late_ph_drop = midpoint["ph"] - final["ph"]
    stalled = abs(late_ph_drop) < STALL_PH_DELTA

    # Stoichiometry: FeS2 gives 2 S per Fe. A ratio far from 2 means dissolved Fe
    # no longer tracks how much pyrite oxidised.
    ratio = None
    if final["fe_mol_kgw"] > 0.0:
        ratio = final["sulfate_mol_kgw"] / final["fe_mol_kgw"]
    stoichiometry_ok = (
        ratio is not None
        and abs(ratio - STOICHIOMETRIC_S_PER_FE) <= STOICHIOMETRY_TOLERANCE
    )

    depleted = consumed_fraction is not None and consumed_fraction >= DEPLETION_FRACTION

    return {
        "oxygen_replenished": spec["replenish_o2"],
        "oxygen_limited": bool(stalled and not depleted),
        "late_ph_change": late_ph_drop,
        "pyrite_consumed_fraction": consumed_fraction,
        "pyrite_depleted": bool(depleted),
        "sulfate_to_iron_ratio": ratio,
        "stoichiometry_consistent": bool(stoichiometry_ok),
        "initial_ph": first["ph"],
        "final_ph": final["ph"],
        "simulated_days": final["time_days"],
        "rate_is_laboratory_derived": True,
    }


def main(argv: list) -> int:
    parser = argparse.ArgumentParser(
        description="Simulate pyrite oxidation kinetics from a JSON request on stdin.",
    )
    parser.add_argument("--input", help="path to a JSON request file; omit to read stdin")
    try:
        args = parser.parse_args(argv)
    except SystemExit:
        emit({"status": "invalid_request", "error": "bad arguments"})
        return INVALID_REQUEST

    try:
        if args.input:
            if not os.path.isfile(args.input):
                fail_invalid("input file not found: {}".format(args.input))
            payload = Path(args.input).read_text(encoding="utf-8")
        else:
            payload = sys.stdin.read()
    except OSError as error:
        fail_invalid("could not read request: {}".format(error))

    try:
        request = json.loads(payload)
    except json.JSONDecodeError as error:
        fail_invalid("request is not valid JSON: {}".format(error))
    if not isinstance(request, dict):
        fail_invalid("request must be a JSON object")

    try:
        result = run(request)
    except SystemExit:
        raise
    except Exception:
        fail_error(traceback.format_exc(limit=20))

    emit(result)
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: src/pyrite_kinetics.rs
//! Time-dependent pyrite oxidation (acid mine drainage generation) via PHREEQC
//! KINETICS, executed as a Python subprocess.
//!
//! This closes the last gap between the two AMD tools that already exist:
//! static ABA screening (MPA/NAPP) says how much acid a rock *could* produce,
//! and `phreeqc_speciation` says what a given water *is*. Neither answers how
//! fast acid appears, which is what decides whether a pit needs treatment in
//! month three or year thirty.
//!
//! Rate law: Williamson & Rimstidt (1994), taken from the RATES block of the
//! loaded database rather than reimplemented here.
//!
//! Four honesty guards travel with every result:
//!
//! * `oxygen_limited` — a closed system consumes its dissolved O2 and the
//!   reaction stalls. The pH curve then flattens at a value that looks like a
//!   stable long-term outcome but is purely an artifact of the sealed box.
//! * `pyrite_depleted` — once the sulfide is spent the curve also flattens, for
//!   the opposite and physically real reason. The two must not be confused.
//! * `stoichiometry_consistent` — FeS2 releases 2 mol S per mol Fe. When iron
//!   precipitates as ferrihydrite the ratio explodes (observed: 73,377), so
//!   dissolved Fe stops being a measure of how much pyrite oxidised.
//! * `rate_is_laboratory_derived` — always true, and always reported: field
//!   rates are commonly one to two orders of magnitude slower than laboratory
//!   rates, so an uncalibrated absolute timescale is not a prediction.

use serde::{Deserialize, Serialize};
use std::path::PathBuf;
use std::process::Stdio;
use std::time::Duration;

use crate::swmm_runner::swmm_python;

/// Absolute path of the Python wrapper, resolved at compile time.
pub const PYRITE_SCRIPT: &str =
    concat!(env!("CARGO_MANIFEST_DIR"), "/scripts/pyrite_kinetics.py");

/// Hard cap on captured stdout: 8 MiB.
pub const MAX_STDOUT_BYTES: usize = 8 * 1024 * 1024;

/// Resolve the wrapper script, honouring an environment override.
pub fn pyrite_script() -> String {
    std::env::var("ENV_INDONESIA_PYRITE_SCRIPT").unwrap_or_else(|_| PYRITE_SCRIPT.to_string())
}

/// The interpreter is the same venv that carries `phreeqpython`.
pub fn pyrite_python() -> String {
    swmm_python()
}

#[derive(Debug, Clone, Serialize, Deserialize, schemars::JsonSchema)]
pub struct PyriteKineticsRequest {
    /// Reactive pyrite available per kg of pore water, mol/kgw.
    pub pyrite_mol_per_kgw: f64,
    #[serde(default)]
    pub initial_ph: Option<f64>,
    #[serde(default)]
    pub initial_o2_mmol: Option<f64>,
    /// When true the system stays in contact with the atmosphere (open pit or
    /// waste dump). When false the dissolved O2 is finite and the reaction will
    /// stall, which is reported as `oxygen_limited`.
    #[serde(default = "default_true")]
    pub replenish_o2: bool,
    #[serde(default)]
    pub o2_partial_pressure_log10: Option<f64>,
    /// Cumulative, strictly increasing output times in days.
    #[serde(default)]
    pub steps_days: Option<Vec<f64>>,
    /// Williamson & Rimstidt parameters: [log10(A/V), exp(m/m0), exp(O2), exp(H+)].
    #[serde(default)]
    pub parms: Option<Vec<f64>>,
    #[serde(default)]
    pub temperature_c: Option<f64>,
    /// Buffering / precipitating phases held at equilibrium, e.g. `["Calcite"]`.
    #[serde(default)]
    pub neutralising_phases: Option<Vec<String>>,
    #[serde(default)]
    pub timeout_secs: Option<u64>,
}

fn default_true() -> bool {
    true
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KineticStep {
    pub time_days: f64,
    pub ph: f64,
    #[serde(default)]
    pub pe: Option<f64>,
    pub fe_mol_kgw: f64,
    pub sulfate_mol_kgw: f64,
    #[serde(default)]
    pub pyrite_remaining_mol_kgw: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KineticGuards {
    pub oxygen_replenished: bool,
    pub oxygen_limited: bool,
    pub late_ph_change: f64,
    #[serde(default)]
    pub pyrite_consumed_fraction: Option<f64>,
    pub pyrite_depleted: bool,
    #[serde(default)]
    pub sulfate_to_iron_ratio: Option<f64>,
    pub stoichiometry_consistent: bool,
    pub initial_ph: f64,
    pub final_ph: f64,
    pub simulated_days: f64,
    pub rate_is_laboratory_derived: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PyriteKineticsResult {
    pub status: String,
    pub database: String,
    pub database_sha256: String,
    pub rate_law: String,
    pub parms: Vec<f64>,
    pub replenish_o2: bool,
    #[serde(default)]
    pub neutralising_phases: Vec<String>,
    pub initial_pyrite_mol_kgw: f64,
    pub series: Vec<KineticStep>,
    pub guards: KineticGuards,
}

#[derive(Debug, Clone, Deserialize)]
struct PyriteFailure {
    status: String,
    #[serde(default)]
    error: String,
}

/// Validate the request before spending a subprocess on it.
pub fn validate_request(request: &PyriteKineticsRequest) -> Result<(), String> {
    if !request.pyrite_mol_per_kgw.is_finite() || request.pyrite_mol_per_kgw <= 0.0 {
        return Err("pyrite_mol_per_kgw must be a positive finite number".to_string());
    }
    if let Some(ph) = request.initial_ph {
        if !ph.is_finite() || !(0.0..14.0).contains(&ph) {
            return Err(format!("initial_ph must be between 0 and 14 (got {})", ph));
        }
    }
    if let Some(o2) = request.initial_o2_mmol {
        if !o2.is_finite() || o2 < 0.0 {
            return Err("initial_o2_mmol must not be negative".to_string());
        }
    }
    if let Some(temperature) = request.temperature_c {
        if !temperature.is_finite() || !(-10.0..=100.0).contains(&temperature) {
            return Err(format!(
                "temperature_c must be between -10 and 100 (got {})",
                temperature
            ));
        }
    }
    if let Some(parms) = &request.parms {
        if parms.len() != 4 {
            return Err(format!(
                "parms must contain exactly 4 numbers, got {}",
                parms.len()
            ));
        }
        if parms.iter().any(|value| !value.is_finite()) {
            return Err("parms must contain only finite numbers".to_string());
        }
    }
    if let Some(steps) = &request.steps_days {
        if steps.is_empty() {
            return Err("steps_days must not be empty".to_string());
        }
        let mut previous = 0.0_f64;
        for (index, step) in steps.iter().enumerate() {
            if !step.is_finite() || *step <= 0.0 {
                return Err(format!("steps_days[{}] must be positive", index));
            }
            if *step <= previous {
                return Err(
                    "steps_days must be strictly increasing cumulative times".to_string()
                );
            }
            previous = *step;
        }
    }
    Ok(())
}

/// True when the simulated acid-generation trajectory can be read as a
/// screening-level trend rather than an artifact.
///
/// A run that stalled because its sealed box ran out of oxygen, or whose
/// dissolved iron no longer tracks oxidation, describes the model setup rather
/// than the waste rock.
pub fn trajectory_is_interpretable(result: &PyriteKineticsResult) -> bool {
    !result.guards.oxygen_limited && result.guards.stoichiometry_consistent
}

/// Run a pyrite oxidation kinetics simulation.
pub async fn run_pyrite_kinetics(
    request: &PyriteKineticsRequest,
) -> Result<PyriteKineticsResult, String> {
    validate_request(request)?;
    let timeout_secs = request.timeout_secs.unwrap_or(180);
    if timeout_secs == 0 {
        return Err("pyrite kinetics timeout_secs must be greater than zero".to_string());
    }

    let script = pyrite_script();
    if !PathBuf::from(&script).is_file() {
        return Err(format!("pyrite kinetics wrapper script not found: {}", script));
    }
    let payload = serde_json::to_vec(request)
        .map_err(|error| format!("could not serialise pyrite kinetics request: {}", error))?;

    let python = pyrite_python();
    let mut command = tokio::process::Command::new(&python);
    command
        .arg(&script)
        .stdin(Stdio::piped())
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .kill_on_drop(true);

    let mut child = command.spawn().map_err(|error| {
        format!("Failed to spawn pyrite kinetics runner ({}): {}", python, error)
    })?;

    if let Some(mut stdin) = child.stdin.take() {
        use tokio::io::AsyncWriteExt;
        stdin
            .write_all(&payload)
            .await
            .map_err(|error| format!("could not write pyrite kinetics request: {}", error))?;
        stdin
            .shutdown()
            .await
            .map_err(|error| format!("could not close pyrite kinetics stdin: {}", error))?;
    }

    let output = tokio::time::timeout(Duration::from_secs(timeout_secs), child.wait_with_output())
        .await
        .map_err(|_| format!("pyrite kinetics run exceeded {}s timeout", timeout_secs))?
        .map_err(|error| format!("pyrite kinetics runner failed: {}", error))?;

    if output.stdout.len() > MAX_STDOUT_BYTES {
        return Err(format!(
            "pyrite kinetics runner stdout exceeded {} bytes",
            MAX_STDOUT_BYTES
        ));
    }

    let stdout = String::from_utf8_lossy(&output.stdout);
    let line = stdout
        .lines()
        .find(|line| !line.trim().is_empty())
        .unwrap_or("")
        .trim();

    if line.is_empty() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(format!(
            "pyrite kinetics runner produced no JSON (exit {:?}): {}",
            output.status.code(),
            stderr.trim()
        ));
    }

    if !output.status.success() {
        let detail = serde_json::from_str::<PyriteFailure>(line)
            .map(|failure| format!("{}: {}", failure.status, failure.error))
            .unwrap_or_else(|_| line.to_string());
        return Err(format!(
            "pyrite kinetics runner exited with {:?}: {}",
            output.status.code(),
            detail
        ));
    }

    let result: PyriteKineticsResult = serde_json::from_str(line)
        .map_err(|error| format!("pyrite kinetics runner emitted unparseable JSON: {}", error))?;

    if result.status != "ok" {
        return Err(format!("pyrite kinetics run status was {}", result.status));
    }
    if result.series.is_empty() {
        return Err("pyrite kinetics run returned an empty time series".to_string());
    }

    Ok(result)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn valid_request() -> PyriteKineticsRequest {
        PyriteKineticsRequest {
            pyrite_mol_per_kgw: 0.05,
            initial_ph: Some(6.5),
            initial_o2_mmol: Some(0.27),
            replenish_o2: true,
            o2_partial_pressure_log10: None,
            steps_days: Some(vec![1.0, 30.0, 90.0, 180.0, 365.0]),
            parms: None,
            temperature_c: None,
            neutralising_phases: None,
            timeout_secs: Some(180),
        }
    }

    fn result_with(oxygen_limited: bool, stoichiometry_consistent: bool) -> PyriteKineticsResult {
        PyriteKineticsResult {
            status: "ok".to_string(),
            database: "wateq4f_PWN_repaired.dat".to_string(),
            database_sha256: "c0f6".to_string(),
            rate_law: "Williamson & Rimstidt (1994)".to_string(),
            parms: vec![1.0, 0.67, 0.5, -0.11],
            replenish_o2: !oxygen_limited,
            neutralising_phases: vec![],
            initial_pyrite_mol_kgw: 0.05,
            series: vec![KineticStep {
                time_days: 365.0,
                ph: 2.99,
                pe: Some(13.0),
                fe_mol_kgw: 7.57e-4,
                sulfate_mol_kgw: 1.51e-3,
                pyrite_remaining_mol_kgw: Some(0.04924),
            }],
            guards: KineticGuards {
                oxygen_replenished: !oxygen_limited,
                oxygen_limited,
                late_ph_change: if oxygen_limited { -3e-5 } else { 0.381 },
                pyrite_consumed_fraction: Some(0.0151),
                pyrite_depleted: false,
                sulfate_to_iron_ratio: Some(if stoichiometry_consistent { 2.0 } else { 73377.0 }),
                stoichiometry_consistent,
                initial_ph: 4.97,
                final_ph: 2.99,
                simulated_days: 365.0,
                rate_is_laboratory_derived: true,
            },
        }
    }

    #[test]
    fn accepts_a_well_formed_request() {
        assert!(validate_request(&valid_request()).is_ok());
    }

    #[test]
    fn non_positive_pyrite_is_rejected() {
        let mut request = valid_request();
        request.pyrite_mol_per_kgw = 0.0;
        assert!(validate_request(&request)
            .unwrap_err()
            .contains("pyrite_mol_per_kgw"));

        let mut nan = valid_request();
        nan.pyrite_mol_per_kgw = f64::NAN;
        assert!(validate_request(&nan).is_err());
    }

    #[test]
    fn out_of_range_ph_and_temperature_are_rejected() {
        let mut bad_ph = valid_request();
        bad_ph.initial_ph = Some(20.0);
        assert!(validate_request(&bad_ph).unwrap_err().contains("initial_ph"));

        let mut bad_temp = valid_request();
        bad_temp.temperature_c = Some(500.0);
        assert!(validate_request(&bad_temp)
            .unwrap_err()
            .contains("temperature_c"));
    }

    #[test]
    fn steps_must_be_strictly_increasing() {
        let mut request = valid_request();
        request.steps_days = Some(vec![30.0, 30.0]);
        assert!(validate_request(&request)
            .unwrap_err()
            .contains("strictly increasing"));

        let mut backwards = valid_request();
        backwards.steps_days = Some(vec![90.0, 30.0]);
        assert!(validate_request(&backwards).is_err());

        let mut empty = valid_request();
        empty.steps_days = Some(vec![]);
        assert!(validate_request(&empty).unwrap_err().contains("empty"));
    }

    #[test]
    fn parms_must_have_exactly_four_entries() {
        let mut request = valid_request();
        request.parms = Some(vec![1.0, 0.67]);
        assert!(validate_request(&request).unwrap_err().contains("exactly 4"));
    }

    #[test]
    fn oxygen_limited_run_is_not_interpretable() {
        // A sealed system exhausts its dissolved O2 and the pH curve flattens.
        // That plateau describes the box, not the waste rock.
        let stalled = result_with(true, true);
        assert!(stalled.guards.oxygen_limited);
        assert!(!trajectory_is_interpretable(&stalled));
    }

    #[test]
    fn iron_precipitation_breaks_the_stoichiometry_guard() {
        // FeS2 gives 2 mol S per mol Fe. When ferrihydrite precipitates the
        // observed ratio reached 73,377, so dissolved Fe no longer measures how
        // much pyrite oxidised.
        let precipitating = result_with(false, false);
        assert!(precipitating.guards.sulfate_to_iron_ratio.unwrap() > 100.0);
        assert!(!precipitating.guards.stoichiometry_consistent);
        assert!(!trajectory_is_interpretable(&precipitating));
    }

    #[test]
    fn healthy_open_system_run_is_interpretable() {
        let healthy = result_with(false, true);
        assert!(trajectory_is_interpretable(&healthy));
        // The laboratory-rate caveat is never dropped, even for a good run.
        assert!(healthy.guards.rate_is_laboratory_derived);
    }

    #[test]
    fn parses_the_documented_contract() {
        let json = r#"{"status":"ok","database":"wateq4f_PWN_repaired.dat","database_sha256":"c0f6","rate_law":"Williamson & Rimstidt (1994)","parms":[1.0,0.67,0.5,-0.11],"replenish_o2":true,"neutralising_phases":["Calcite"],"initial_pyrite_mol_kgw":0.05,"series":[{"time_days":1.0,"ph":4.969,"pe":13.2,"fe_mol_kgw":3.528e-06,"sulfate_mol_kgw":7.056e-06,"pyrite_remaining_mol_kgw":0.05},{"time_days":365.0,"ph":2.986,"pe":13.4,"fe_mol_kgw":7.572e-04,"sulfate_mol_kgw":1.5144e-03,"pyrite_remaining_mol_kgw":0.04924}],"guards":{"oxygen_replenished":true,"oxygen_limited":false,"late_ph_change":0.381,"pyrite_consumed_fraction":0.0151,"pyrite_depleted":false,"sulfate_to_iron_ratio":2.0,"stoichiometry_consistent":true,"initial_ph":4.969,"final_ph":2.986,"simulated_days":365.0,"rate_is_laboratory_derived":true}}"#;
        let parsed: PyriteKineticsResult = serde_json::from_str(json).unwrap();
        assert_eq!(parsed.series.len(), 2);
        assert_eq!(parsed.neutralising_phases, vec!["Calcite".to_string()]);
        assert!((parsed.guards.final_ph - 2.986).abs() < 1e-9);
        assert!(trajectory_is_interpretable(&parsed));
    }

    #[test]
    fn wrapper_script_path_is_absolute_and_present() {
        let script = std::path::Path::new(PYRITE_SCRIPT);
        assert!(script.is_absolute(), "PYRITE_SCRIPT must not depend on CWD");
        assert!(script.is_file(), "missing wrapper script at {}", PYRITE_SCRIPT);
    }

    #[tokio::test]
    async fn zero_timeout_is_rejected_before_spawning() {
        let mut request = valid_request();
        request.timeout_secs = Some(0);
        let error = run_pyrite_kinetics(&request).await.unwrap_err();
        assert!(error.contains("timeout_secs"), "unexpected: {error}");
    }
}
